File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0

        result = 0
        queue = deque([(root, float("-inf"))])

        while queue:
            node, maxVal = queue.popleft()
            if node.val >= maxVal:
                result += 1
            maxVal = max(maxVal, node.val)
            if node.left:
                queue.append([node.left, maxVal])
            if node.right:
                queue.append([node.right, maxVal])

        return result

        

File: Data Structures & Algorithms/max-water-container/submission-3.java
class Solution {
    public int maxArea(int[] heights) {
        int result = 0;
        int n = heights.length;
        int left = 0, right = n - 1;

        while (left < right) {
            int lowwerHeight = Math.min(heights[left], heights[right]);
            result = Math.max(result, (right - left) * lowwerHeight);
            if (heights[left] < heights[right]) {
                left++;
            } else {
                right--;
            }
        }
        return result;
    }
}
